### PR TITLE
feat: add Wave Function Collapse

### PR DESCRIPTION
**Pull request by Tboefijn**
## Description

The wave collapse function is responsible for level generation. It needs to:
- Generate a level based on a grid.
  - Make sure the grid is variable (you don’t have to implement the configuration system now, but make it variable in the function)
- Follow the research done by @SSelinn 

## Testability

Open the Wave Function Collapse scene. If you select `Grid` in the hierarchy you can edit some parameters in the Inspector.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 1]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
         /// The size of the tiles in the y-direction.
         /// </summary>
-        private const int _tileSizeY = 16;
+        [FormerlySerializedAs("tileSizeY")] public int tileSizeZ = 16;
 
         /// <summary>
         /// Represents the room objects.
         /// </summary>
-        public RoomObjects RoomObjects;
+        public RoomObjects roomObjects;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:33]:
```suggestion
        [SerializeField]
        private RoomObjects _roomObjects;
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 2]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+
+        /// <summary>
+        /// A boolean that indicates whether a seed is used.
+        /// </summary>
+        public bool useSeed;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:38]:
```suggestion
        [SerializeField]
        private bool _useSeed;
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 3]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+
+        /// <summary>
+        /// The seed used for the random number generator.
+        /// </summary>
+        public int seed;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:43]:
```suggestion
        [SerializeField]
        private int _seed;
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 4]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+
+        /// <summary>
+        /// The random number generator.
+        /// </summary>
+        private Random _random = new();
+
+        /// <summary>
+        /// The width of the grid in the x-direction.
+        /// </summary>
+        public int gridWidthX = 10;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:53]:
```suggestion
        [SerializeField]
        private int _gridWidthX = 10;
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 5]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+
+        /// <summary>
+        /// The width of the grid in the z-direction.
+        /// </summary>
+        public int gridWidthZ = 10;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:58]:
```suggestion
        [SerializeField]
        private int _gridWidthZ = 10;
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 6]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+
+        /// <summary>
+        /// The amount of rooms that need to be placed.
+        /// </summary>
+        public int amountOfRooms = 5;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:63]:
```suggestion
        [SerializeField]
        private int _amountOfRooms = 5;
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 7]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+        {
+            if (amountOfRooms > gridWidthX * gridWidthZ)
+            {
+                throw new Exception("The amount of rooms is larger than the available places in the grid.");
+            }
+
+            if (useSeed)
+            {
+                _random = new Random(seed);
+            }
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:87]:
```suggestion
            if (amountOfRooms > gridWidthX * gridWidthZ)
                throw new Exception("The amount of rooms is larger than the available places in the grid.");

            if (useSeed) _random = new Random(seed);
```

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 8]
```diff
@@ -18,90 +20,147 @@ public class GridPlacer : MonoBehaviour
...
+        /// Makes the grid.
+        /// </summary>
+        private void MakeGrid()
+        {
+            _grid = new Grid(gridWidthX, gridWidthZ, _random);
 
             _grid.Init();
 
-            TempFillFunction();
+            int noRooms = 0;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:105]:
```suggestion
            int numberOfRooms = 0;
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Awesome work! Left some style comments, but the generation looks amazing!

File: aplib.net-demo/Assets/Scripts/WFC/Grid.cs [hunk 9]
```diff
@@ -80,7 +100,7 @@ public void Init()
         /// <exception cref="IndexOutOfRangeException">
         /// When index is larger than <see cref="Width"/> times <see cref="Height"/>.
         /// </exception>
-        protected (int x, int y) IndexToCoordinates(int index)
+        private (int x, int z) IndexToCoordinates(int index)
```
[the remark this answers is not in the document]
Comment (Tboefijn) [reply aplib.net-demo/Assets/Scripts/WFC/Grid.cs:103]: In de toekomst komt hier wellicht wel een subclass van dus ik laat het voor nu even zo staan.

**Review by JensSteenmetz — APPROVED** [review]

**Comment by JensSteenmetz** [comment]
I forgor: LGTM

**Pull request merged**